File: src/MotorNode.py
#!/usr/bin/env python
from PololuSMC import MotorController
import time
import rospy
from std_msgs.msg import Bool
from std_msgs.msg import Float32MultiArray
from std_msgs.msg import UInt16MultiArray
from std_msgs.msg import Int16MultiArray

port_name = "/dev/ttyS0"
baud_rate = 9600
max_speed = 1

def vel_callback(data):
    global controller
    for motor in range(4):
        speed = int(round(data.data[motor]/max_speed * 3200))
        controller.set_target_speed(motor, speed)

def start_callback(data):
    global controller
    if(data.data):
        rospy.loginfo("Starting all motors")
        controller.safe_start_all()
    else:
        rospy.loginfo("Stopping all motors")
        controller.stop_all()

def brake_callback(data):
    global controller
    if(data.data):
        rospy.loginfo("Braking all motors")
        controller.brake_all()
    
def init_control():
    global controller

    rospy.init_node('motor_control')

    rospy.Subscriber("/motor_vel", Float32MultiArray, vel_callback)
    rospy.Subscriber("/motor_cmd/start", Bool, start_callback)
    rospy.Subscriber("/motor_cmd/brake", Bool, brake_callback)

    error_pub = rospy.Publisher("/motor_status/errors", UInt16MultiArray, queue_size=10)
    limit_pub = rospy.Publisher("/motor_status/limit_status", UInt16MultiArray, queue_size=10)
    target_pub = rospy.Publisher("/motor_status/target_speeds", Int16MultiArray, queue_size=10)
    speed_pub = rospy.Publisher("/motor_status/current_speeds", Int16MultiArray, queue_size=10)
    temp_pub = rospy.Publisher("/motor_status/temps", UInt16MultiArray, queue_size=10)
    curr_pub = rospy.Publisher("/motor_status/currents", UInt16MultiArray, queue_size=10)
    volt_pub = rospy.Publisher("/motor_status/voltages", UInt16MultiArray, queue_size=10)

    controller = MotorController(port_name, baud_rate, debug=True)

    rate = rospy.Rate(10)
    cycle = 10
    while not rospy.core.is_shutdown():

        if(cycle % 1 == 0):
            pub_speeds(speed_pub)
        if((cycle-1) % 30 == 0):
            pub_errors(error_pub)
        if((cycle-2) % 20 == 0):
            pub_limits(limit_pub)
        if((cycle-3) % 1000 == 0):
            pub_targets(target_pub)
        if((cycle-4) % 50 == 0):
            pub_temps(temp_pub)
        if((cycle-5) % 5 == 0):
            pub_curr(curr_pub)
        if((cycle-6) % 5 == 0):
            pub_volt(volt_pub)

        cycle += 1
        if(cycle > 500):
            cycle = cycle % 500
        rate.sleep()

def pub_errors(pub):
    status = [0] * 4
    for motor in range(4):
        status[motor] = controller.get_error_status(motor)

    array = UInt16MultiArray()
    array.data = status
    pub.publish(array) 

def pub_limits(pub):
    status = [0] * 4
    for motor in range(4):
        status[motor] = controller.get_limit_status(motor)

    array = UInt16MultiArray()
    array.data = status
    pub.publish(array) 

def pub_targets(pub):
    speeds = [0] * 4
    for motor in range(4):
        speeds[motor] = controller.get_target_speed(motor)

    array = Int16MultiArray()
    array.data = speeds
    pub.publish(array) 

def pub_speeds(pub):
    speeds = [0] * 4
    for motor in range(4):
        speeds[motor] = controller.get_current_speed(motor)

    array = Int16MultiArray()
    array.data = speeds
    pub.publish(array) 

def pub_temps(pub):
    temps = [0] * 4
    for motor in range(4):
        (temp_A, temp_B) = controller.get_temperatures(motor)
        temps[motor] =(temp_A + temp_B)/2

    array = UInt16MultiArray()
    array.data = temps
    pub.publish(array) 

def pub_curr(pub):
    curr = [0] * 4
    for motor in range(4):
        curr[motor] = controller.get_current(motor)

    array = UInt16MultiArray()
    array.data = curr
    pub.publish(array) 

def pub_volt(pub):
    volt = [0] * 4
    for motor in range(4):
        volt[motor] = controller.get_input_voltage(motor)

    array = UInt16MultiArray()
    array.data = volt
    pub.publish(array) 

if __name__ == '__main__':
    init_control()

File: src/MotorNode.cpp
#include "PololuSMC.hpp"
#include <vector>
#include <math.h>

#include "ros/ros.h"

#include "std_msgs/Bool.h"
#include "std_msgs/Float32.h"
#include "std_msgs/Float32MultiArray.h"
#include "std_msgs/UInt16MultiArray.h"
#include "std_msgs/Int16MultiArray.h"

using namespace std;

PololuSMC controller;

ros::Publisher error_pub;
ros::Publisher limit_pub;
ros::Publisher target_pub;
ros::Publisher speed_pub;
ros::Publisher temp_pub;
ros::Publisher curr_pub;
ros::Publisher volt_pub;

float max_speed = 1;
void vel_callback(const std_msgs::Float32MultiArray::ConstPtr& msg){
    for(int i=0; i < 4; i++){
        int16_t speed = (int16_t) round(msg->data[i]*max_speed*1600);
        controller.set_target_speed(i, speed);
    }
}
void tray_callback(const std_msgs::Float32::ConstPtr& msg){
    int16_t speed = (int16_t) round(msg->data*max_speed*3200);
    controller.set_target_speed(4, speed);
}
void start_callback(const std_msgs::Bool::ConstPtr& msg){
    if(msg->data){
        ROS_INFO("Starting all motors");
        controller.safe_start_all();
    }
    else{
        ROS_INFO("Stopping all motors");
        controller.stop_all();
    }
}
void brake_callback(const std_msgs::Bool::ConstPtr& msg){
    if(msg->data){
        ROS_INFO("Braking all motors");
        controller.brake_all();
    }
}
void pub_errors(){
    vector<uint16_t> status(5);
    for(int i = 0; i < 5; i++){
        status[i] = controller.get_error_status(i);
    }
    std_msgs::UInt16MultiArray msg;
    msg.data=status;
    error_pub.publish(msg);
}
void pub_limits(){
    vector<uint16_t> status(5);
    for(int i = 0; i < 5; i++){
        status[i] = controller.get_limit_status(i);
    }
    std_msgs::UInt16MultiArray msg;
    msg.data=status;
    limit_pub.publish(msg);
}
void pub_targets(){
    vector<int16_t> status(5);
    for(int i = 0; i < 5; i++){
        status[i] = controller.get_target_speed(i);
    }
    std_msgs::Int16MultiArray msg;
    msg.data=status;
    target_pub.publish(msg);
}
void pub_speeds(){
    vector<int16_t> status(5);
    for(int i = 0; i < 5; i++){
        status[i] = controller.get_current_speed(i);
    }
    std_msgs::Int16MultiArray msg;
    msg.data=status;
    speed_pub.publish(msg);
}
void pub_temps(){
    vector<uint16_t> status(5);
    for(int i = 0; i < 5; i++){
        status[i] = controller.get_temperatures(i);
    }
    std_msgs::UInt16MultiArray msg;
    msg.data=status;
    temp_pub.publish(msg);
}
void pub_curr(){
    vector<uint16_t> status(5);
    for(int i = 0; i < 5; i++){
        status[i] = controller.get_current(i);
    }
    std_msgs::UInt16MultiArray msg;
    msg.data=status;
    temp_pub.publish(msg);
}
void pub_volt(){
    vector<uint16_t> status(5);
    for(int i = 0; i < 5; i++){
        status[i] = controller.get_input_voltage(i);
    }
    std_msgs::UInt16MultiArray msg;
    msg.data=status;
    volt_pub.publish(msg);
}

int main(int argc, char **argv){
    ros::init(argc, argv, "motornode");
    ros::NodeHandle nh;

    ROS_INFO("Motor Node: Starting serial controller...");
    controller.init_port();
    printf("Stopping all motors...");
    controller.stop_all();
    printf("Success!\n");
    
    printf("Initializing subscribers and publishers...");
    ros::Subscriber vel_sub = nh.subscribe("/motor_vel", 1, vel_callback);
    ros::Subscriber tray_sub = nh.subscribe("/tray_vel", 1, tray_callback);
    ros::Subscriber start_sub = nh.subscribe("/motor_cmd/start", 1, start_callback);
    ros::Subscriber brake_sub = nh.subscribe("/motor_cmd/brake", 1, brake_callback);

    error_pub = nh.advertise<std_msgs::UInt16MultiArray>("/motor_status/errors",1);
    limit_pub = nh.advertise<std_msgs::UInt16MultiArray>("/motor_status/limit_status",1);
    target_pub = nh.advertise<std_msgs::Int16MultiArray>("/motor_status/target_speeds",1);
    speed_pub = nh.advertise<std_msgs::Int16MultiArray>("/motor_status/current_speeds",1);
    temp_pub = nh.advertise<std_msgs::UInt16MultiArray>("/motor_status/temps",1);
    curr_pub = nh.advertise<std_msgs::UInt16MultiArray>("/motor_status/currents",1);
    volt_pub = nh.advertise<std_msgs::UInt16MultiArray>("/motor_status/voltages",1);
    printf("Success!\n");

    ros::Rate rate(1000);
    int cycle = 10;

    printf("Starting publishing loop...\n");
    while(ros::ok()){
        ros::spinOnce();
        if(cycle % 1 == 0){
            pub_speeds();
        }
        if((cycle-1) % 2 == 0){
            pub_errors();
        }
        if((cycle-2) % 100 == 0){
            pub_limits();
        }
        if((cycle-3) % 1 == 0){
            pub_targets();
        }
        if((cycle-4) % 50 == 0){
            pub_temps();
        }
        if((cycle-5) % 5 == 0){
            pub_curr();
        }
        if((cycle-6) % 5 == 0){
            pub_volt();
        }
        cycle += 1;
        if(cycle > 500){
            cycle = cycle % 500;
        }
        rate.sleep();
    }
    ROS_INFO("Motor Node exiting...");
    return 0;
}

